libs/providers/multi-provider/src/lib/multi-provider.ts

```diff
@@ -0,0 +1,402 @@
+import {
+  DefaultLogger,
+  EvaluationContext,
+  EventDetails,
+  FlagValueType,
+  GeneralError,
+  Hook,
+  HookContext,
+  HookHints,
+  JsonValue,
+  Logger,
+  OpenFeatureEventEmitter,
+  Provider,
+  ProviderEvents,
+  ProviderStatus,
+  ResolutionDetails,
+} from '@openfeature/server-sdk';
+import { BeforeHookContext, ProviderMetadata } from '@openfeature/core';
+import { BaseEvaluationStrategy, FirstMatchStrategy, ProviderResolutionResult } from './strategies';
+import { HookExecutor } from './hook-executor';
+import { AggregateError } from './errors';
+
+// Represents an entry in the constructor's provider array which may or may not have a name set
+export type ProviderEntryInput = {
+  provider: Provider;
+  name?: string;
+};
+
+// Represents a processed and "registered" provider entry where a name has been chosen
+type RegisteredProvider = {
+  name: string;
+  provider: Provider;
+};
+
+export class MultiProvider implements Provider {
+  readonly runsOn = 'server';
+
+  public readonly events = new OpenFeatureEventEmitter();
+
+  private hookContexts: WeakMap<EvaluationContext, HookContext> = new WeakMap<EvaluationContext, HookContext>();
+  private hookHints: WeakMap<EvaluationContext, HookHints> = new WeakMap<EvaluationContext, HookHints>();
+
+  metadata: ProviderMetadata;
+
+  providerEntries: RegisteredProvider[] = [];
+  private providerEntriesByName: Record<string, RegisteredProvider> = {};
+  private readonly providerStatuses: Record<string, ProviderStatus> = {};
+
+  private hookExecutor: HookExecutor;
+
+  constructor(
+    readonly constructorProviders: ProviderEntryInput[],
+    private readonly evaluationStrategy: BaseEvaluationStrategy = new FirstMatchStrategy(),
+    private readonly logger: Logger = new DefaultLogger(),
+  ) {
+    this.hookExecutor = new HookExecutor(this.logger);
+
+    this.registerProviders(constructorProviders);
+
+    // just make sure we don't accidentally modify these later
+    Object.freeze(this.providerEntries);
+    Object.freeze(this.providerEntriesByName);
+
+    this.providerStatuses = {};
+
+    const aggregateMetadata = Object.keys(this.providerEntriesByName).reduce((acc, name) => {
+      return { ...acc, [name]: this.providerEntriesByName[name].provider.metadata };
+    }, {});
+
+    this.metadata = {
+      ...aggregateMetadata,
+      name: MultiProvider.name,
+    };
+  }
+
+  private registerProviders(constructorProviders: ProviderEntryInput[]) {
+    const uniqueNames = new Set();
```

Comment:
Another option would be to just iterate through the `constructorProviders`, and add them to a `map<Name, Array<Provider>>`, then iterate over that map and any with multiple providers in their array you then can just name it according to its index.